Implementation of `downstream` connection with `braidpool` via stratum

Sansh2356 opened the pull request:
- `Stratum` implementation for downstream to upstream communication.
- `IPC` template consumption and `mining.notify` for the job transfer to the downstream node.
- `Tracker` for job mapping wrt to the `downstream` nodes .
- `Client2Server` for messages such as `mining.submit` and difficulty adjustment according to the valid mined bead.
- `Validation` of job received from the downstream and constructing valid `bead` .

Sansh2356 committed:
Adding stratum implementation

Sansh2356 commented:
CC by @mcelrath  at PR #233 

Regarding difficulty adjustment:

There are TWO difficulties involved here:

    a per-miner difficulty.
    the Braidpool network difficulty.

(1) We will have a measurement of the network latency parameter a from the braid. The per-miner difficulty should be chosen so that each mining device produces a response regularly, (roughly every a (seconds) or some multiple of a). The way ckpool targets difficulty is relevant here. We will want to keep statistics about mining devices using the miner's responses (so we can track individual-device hashrate and problems with devices). Note that this rate is not a constant (because a is not constant, and devices can be over/under clocked).

(2) The network difficulty is the subject of my work in PR https://github.com/braidpool/braidpool/pull/91 . You can think of that work as outputting the parameters a and lambda from network/braid measurements which you can use in (1). Secondly it decides when a share becomes a bead. As discussed on the call today, to start we can set this to a constant for simplicity to start, let's say 1/500 of bitcoin's difficulty. (500*bitcoin_target).

Sansh2356 committed:
Replacing placeholders for coinbase splitting

Signed-off-by: sansh2356 <sansh2356@gmail.com>

Sansh2356 committed:
Latest template caching and forwarding after subscrbing by a new downstream

Sansh2356 committed:
Updating stratum response handling

mcelrath commented on node/src/main.rs:
```diff
@@ -51,9 +52,46 @@ mod init_capnp;
+    //latest available template to be cached for the newest connection until new job is received
+    let mut latest_template = Arc::new(Mutex::new(BlockTemplate::default()));
+    //latest available template merkel branch
+    let mut latest_template_merkel_branch = Arc::new(Mutex::new(Vec::new()));
```
It's spelled "Merkle" after Ralph Merkle.

Sansh2356 committed:
Adding unit tests && small fix for job id generation

mcelrath commented on node/src/ipc.rs:
```diff
@@ -299,10 +299,11 @@ async fn get_template_with_retry(
     const MIN_TEMPLATE_SIZE: usize = 512;
     let config = CoinbaseConfig::default();
     let mut last_template = Vec::new();
+    let mut last_template_merkel_branch: Vec<Vec<u8>> = Vec::new();
```
* merkle

Sansh2356 committed:
Adding automated test for stratum service

mcelrath commented on node/src/template_creator.rs:
```diff
@@ -297,8 +299,11 @@ fn build_coinbase_input(
+    //extranonce starts with 8 assigned to extranonce bytes
+    //therefore the extranonce separator must be before this
+    //TODO KINDLY REVERT IF NOT WORKS
+    script_data.push(EXTRANONCE_SEPARATOR.len() as u8);
```
As far as I know there's no such thing as an "extranonce separator" -- there's no separator character. The extranonce_prefix, extranonce, and extranonce_suffix are simply concatenated.

But let me know if I"m wrong...

Sansh2356 replied:
The `extranonce separator` is only for the purpose of distinguishing between the `enonce1` i.e. the prefix and `enonce2` i.e. the rollable suffix bits available to miner. So it helps in `coinbase splitting` while the notification of the job is being sent to the downstream miner .

mcelrath commented on node/src/lib.rs:
```diff
@@ -15,7 +26,105 @@ pub mod peer_manager;
+            log::info!("Transactions count is - {}", template_transactions.len());
+            let template: BlockTemplate = BlockTemplate::default();
+            let mut latest_template = latest_template_arc.lock().await;
+            latest_template.version = template.version;
```
Everything in this section should be using `candidate_block` instead of `template`.

Sansh2356 replied:
The `template` itself is being de-serialized into `candidate_block` hence the job is being constructed according to the candidate block only .

mcelrath commented on node/src/stratum.rs:
```diff
@@ -0,0 +1,1637 @@
+        log::info!("Subscribing is taking place -- {:?}", subscribe_req_params);
+        //FIXME dummy testing subscription IDs must be unique though
+        let subscriptions: Vec<(String, String)> = vec![
+            (String::from("mining.set_difficulty"), String::from("34")),
```
Can you explain these constants "34" and "12"? I think one of them should be a UUID, no?

Sansh2356 replied:
There is no specific reason for this, separate channel ID's are required for both the `subscribe` and `notify` channels so i have allotted them randomly so they are just placeholders can be changed accordingly .

mcelrath commented on node/src/stratum.rs:
```diff
@@ -0,0 +1,1637 @@
+        mut mining_job_map: Arc<Mutex<MiningJobMap>>,
+        client_request_id: u64,
+    ) -> Result<StratumResponses, StratumErrors> {
+        let param_array = submit_work_params.as_array().unwrap();
```
Can cause a panic and crash the node.

There are several unsafe unwrap()'s in this function and handle_configure.

Sansh2356 replied:
ACK .

mcelrath commented on node/src/stratum.rs:
```diff
@@ -0,0 +1,1637 @@
+                    //We will receive the template from the IPC channel and construct a valid job
+                    //from the provided template and pass onto the message_reciver in the handle connection for
+                    // downstream communication to take place.
+                    for (peer_adr, mining_job_arc) in self.job_map_arc.lock().await.iter() {
```
This holds this lock for the entire duration of notifying all clients. Clone the job_map instead.

mcelrath commented on node/src/stratum.rs:
```diff
@@ -0,0 +1,1637 @@
+                                return Err(error);
+                            }
+                        };
+                        //Sending the notification for broadcasting it across all the downstream
```
This is a nested loop. You're sending each notification N times to each client. Did you intend to do this?

Sansh2356 replied:
No correctly pointed out ACK thanks ! .

mcelrath reviewed (changes requested):
This is really excellent Ansh, but jeez it's dense and my brain hurts.

Sansh2356 committed:
Cleanup and small fix for stratum service

doc test fix

cargo fmt test

missing-typo for suggest-difficulty

correcting method signature in test in stratum.rs

mcelrath committed:
Merge branch 'dev' into stratum_test

mcelrath reviewed (commented):
This needs some config options and command line options.

mcelrath committed:
Merge branch 'dev' into stratum_test

Sansh2356 committed:
Merge branch 'dev' into stratum_test

Sansh2356 committed:
Updating downstream client struct

Sansh2356 committed:
Merge branch 'dev' into stratum_test

Sansh2356 committed:
Computed masked version for extended configurations in BIP 310

mcelrath commented on node/src/lib.rs:
```diff
@@ -15,7 +27,156 @@ pub mod peer_manager;
+/// **Length of the extranonce prefix (in bytes).**
+///
+/// In Stratum mining, the extranonce is split into two parts:
+/// `EXTRANONCE1` (prefix) and `EXTRANONCE2` (suffix).
```
This isn't correct, the assembled coinbase should be (schematically):

coinbase = prefix + extranonce1 + separator + extranonce2 + suffix

Sansh2356 replied:
It is done like that only see in coinbase reconstruction while constructing notification, this comment talks about `extranonce`  and its splitting to be done in `extranonce1` and `extranonce2` respectively .

mcelrath commented on node/src/lib.rs:
```diff
@@ -15,7 +27,156 @@ pub mod peer_manager;
+///
+/// This is an array of bytes used to clearly delimit the two extranonce parts.
+/// In this testing configuration, the separator length equals
+/// `EXTRANONCE1_SIZE + EXTRANONCE2_SIZE` (8 bytes total),
```
This isn't correct. Both extranonces are generally fields in the scriptSig of the coinbase transaction. Though in principle it can be any part of the coinbase transaction as long as the transaction remains valid after they are modified (OP_RETURNs are also sometimes used)

https://blockstream.info/tx/f3b1939883707d52c19415d8d16a0cee1ef7475da1cf7e8e7d879ce5ed47d7c0?expand

The scriptSig supposed to be the signature for the inputs of a transactoin, but the coinbase has no inputs, so the scriptSig is a free place to put data in the above transaction it is:
```
OP_PUSHBYTES_3 f8ec0d OP_0 OP_PUSHBYTES_4 7b09b568 OP_PUSHBYTES_4 50b58f05 OP_PUSHBYTES_12 88994b6c5511e36d925c7a01 OP_PUSHBYTES_10 636b706f6f6c112f736f OP_FROMALTSTACK OP_3DUP OP_PUSHBYTES_46 <push past end>
```

Here the first field (f8ec0d) is likely extranonce1, and the second (7b09b568) is extranonce2. Not sure what the others are.

We need to write a valid coinbase transaction for braidpool first, and then identify the bytes which can be modified for extranonce1 and extranonce2.

Sansh2356 replied:
Same as below .

mcelrath commented on node/src/lib.rs:
```diff
@@ -15,7 +27,156 @@ pub mod peer_manager;
+            > = deserialize(&template_bytes.0.clone());
+            let merkle_branch_coinbase = template_bytes.1.clone();
+            let (template_header, template_transactions) = candidate_block.unwrap().into_parts();
+            let coinbase_transaction = template_transactions.get(0);
```
Here we should be modifying the coinbase transaction to *add* the extranonce fields in the scripSig. Then by serializing it we can determine the prefix, separator, and suffix.

Sansh2356 replied:
It is done via `extranonce` separator which is added while constructing coinbase after received from `getblocktemplate` from the bitcoin-node via ipc, kindly check in `template_creator.rs` in `build_coinbase_input` . I have serialized saved reconstructed coinbase too from testing these are added to `ScriptSig` after `tip_height` only .
```
    script_data.extend_from_slice(&EXTRANONCE_SEPARATOR);
```

mcelrath commented on node/src/lib.rs:
```diff
@@ -15,7 +27,156 @@ pub mod peer_manager;
+            let template: BlockTemplate = BlockTemplate {
+                version: template_header.version.to_consensus(),
+                previousblockhash: template_header.prev_blockhash.to_string(),
+                transactions: template_transactions.clone(),
```
Once we modify the Coinbase we need to make sure it ends up here, in the list of txs.

Sansh2356 replied:
It is already returned as the first transaction in `template_consumer.rs` .

mcelrath reviewed (changes requested):
Before merging I'd like to see a bead or block mined using the Stratum interface! ;-)

Sorry this one has been so long...

Sansh2356 committed:
Correcting to 4-byte little-endian conversion && mining.submit test

Sansh2356 committed:
Adding merkle root construction via merkle path instead of txs_ids

cargo fmt check

Small refactor
